### PR TITLE
Updating HAL/VM ABI to support indirect bindings (COMPATIBILITY BREAKING).

### PR DESCRIPTION
**Pull request by benvanik**
This bumps the HAL module version to 0.3 and makes a few changes required to support reusable/indirect command buffers.

This is the first breaking change in the HAL ABI in 2024 and hopefully the only one required this year!

Summary of changes:
* `hal.command_buffer.create` now takes a set of allowed queue affinities to match `iree_hal_command_buffer_create`
* `hal.command_buffer.update_buffer` was added and maps to `iree_hal_command_buffer_update_buffer`
* `hal.command_buffer.fill_buffer`, `hal.command_buffer.copy_buffer`, `hal.command_buffer.collective`, and `hal.command_buffer.dispatch.indirect` were updated to take binding table slots in addition to buffer pointers

Progress on #17875.

### CODE REVIEW

**Review by ScottTodd — APPROVED** [review]
LGTM. Timing wise... I think we could land this whenever. We're due for a stable release soon, so we could cut a release _before_ this change then let this soak until the next release?

A few people are working on branches (like https://github.com/iree-org/iree/tree/shared/sdxl_quantized), so they'll want to be careful/intentional about pulling this in.

**Comment by benvanik** [comment]
PTAL - added an `iree_hal_dispatch_flags_t` placeholder in case we need it in the next ~year.

File: compiler/src/iree/compiler/Dialect/HAL/IR/HALAttrs.td [hunk 1]
```diff
@@ -186,6 +186,14 @@ def HAL_DescriptorSetLayoutFlagsAttr :
   let cppNamespace = "::mlir::iree_compiler::IREE::HAL";
 }
 
+def HAL_DispatchFlags_None : I64BitEnumAttrCase<"None", 0x0000>;
+def HAL_DispatchFlagsAttr :
+    I64BitEnumAttr<"DispatchFlags", "valid Dispatch flags", [
+      HAL_DispatchFlags_None,
+    ]> {
+  let cppNamespace = "::mlir::iree_compiler::IREE::HAL";
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/HAL/IR/HALAttrs.td]: nit: why the capitalization in `valid Dispatch flags`?

**Comment by ScottTodd** [comment]
> PTAL - added an `iree_hal_dispatch_flags_t` placeholder in case we need it in the next ~year.

SGTM. Matches the Vulkan API style of adding flags reserved for future use :P

**Review by ScottTodd — APPROVED** [review]
LGTM when CI passes (some crashes compiling test programs?)

File: compiler/src/iree/compiler/Dialect/HAL/IR/HALAttrs.td (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/HAL/IR/HALAttrs.td]: copy/paste :P fixed

**Comment by benvanik** [comment]
(test_onnx failures seem infra-related, merging past them)

**Pull request merged**